### PR TITLE
Add Knowledge Graph module for LLM-powered entity extraction and graph traversal

### LINKED ISSUE #217: Add Knowledge Graph Module for LLM-powered entity extraction and graph traversal
# Knowledge Graph Module — Tracking Issue

## Status: Phase 1 Complete (MVP)

PR #566 merged the initial Knowledge Graph module with:
- Core data model (`Node`, `Edge`, `Graph` with `ujson.Value` properties)
- LLM-powered entity/relationship extraction (free-form)
- Graph traversal engine (BFS, DFS, shortest path, all paths, distance queries)
- JSON file-based persistence
- Comprehensive test suite

### Phase 2: Storage & Graph Engine Integration

The current in-memory + JSON storage is fine for small graphs, but production use cases need pluggable storage backends and integration with graph engines.

- [ ] #650 — **Pluggable GraphStore trait and embedded storage** — Abstract trait with operations (upsert node/edge, query, traverse), plus SQLite-based embedded implementation
- [ ] #653 — **Graph engine integration adapters** — Adapter layer for external graph databases (Neo4j, TinkerPop/Gremlin, Neptune). Engine-agnostic design.

### Phase 3: GraphRAG

This is the highest-value use case — using knowledge graphs to enhance RAG beyond what vector search alone can do.

- [ ] #651 — **GraphRAG — community detection, summaries, and hybrid retrieval** — Leiden clustering, hierarchical community summaries, global/local search, hybrid vector+graph retrieval bridging existing VectorStore/EmbeddingProvider

### Phase 4: Advanced Extraction

- [ ] #652 — **Schema-guided extraction and multi-document graph building** — User-defined entity/relationship schemas, incremental multi-document extraction, entity linking & disambiguation, coreference resolution

### Phase 5: Query & Interaction

- [ ] #654 — **Natural language graph queries and graph-guided QA** — LLM translates questions to graph traversals, graph-guided question answering, PageRank/centrality for entity ranking

## Architecture Principles

1. **Engine-agnostic** — Don't couple to any specific graph database. Design the `GraphStore` trait so it works with embedded storage, Neo4j, Neptune, TinkerPop, etc.
2. **Build on existing llm4s patterns** — Use `Result[A]`, `LLMClient`, `EmbeddingProvider`, `VectorStore`
3. **Incremental value** — Each phase should be independently useful
4. **Hybrid retrieval first** — The vector+graph bridge (Phase 3) is the highest-value target after storage

## Sub-Issues

| # | Title | Phase | Depends On |
|---|-------|-------|------------|
| #650 | Pluggable GraphStore trait and embedded storage | 2 | — |
| #651 | GraphRAG — community detection, summaries, hybrid retrieval | 3 | #650 |
| #652 | Schema-guided extraction and multi-document graph building | 4 | — |
| #653 | Graph engine integration adapters | 2 | #650 |
| #654 | Natural language graph queries and graph-guided QA | 5 | #650 |

### PR DESCRIPTION
**Pull request by dhairya-pandya**
# Add Knowledge Graph Module

## What does this PR do?

This PR adds a Knowledge Graph module to `llm4s` with three core capabilities:

1. **LLM-Powered Extraction** - Extract entities and relationships from text using any LLM provider
2. **Graph Traversal & Queries** - Navigate graphs using BFS, DFS, shortest path, and custom filters
3. **JSON Storage** - Save and load graphs to/from JSON files

**Why this issue?** Knowledge Graphs are increasingly important for Graph RAG and agentic AI workflows. This adds a foundational capability that `llm4s` was missing.

### Quick Example
```scala
// Extract knowledge from text
val generator = new KnowledgeGraphGenerator(llmClient)
val graph = generator.extract(
  text = "Alice works at Google",
  entityTypes = List("Person", "Organization")
)

// Query and traverse
val engine = new GraphEngine(graph)
val path = engine.findShortestPath("alice", "google")

// Persist
val store = new JsonGraphStore(Paths.get("graph.json"))
store.save(graph)
```

### Files Added
- `knowledgegraph/KnowledgeGraph.scala` - Core data structures (Node, Edge, Graph)
- `knowledgegraph/item/KnowledgeGraphGenerator.scala` - LLM extraction
- `knowledgegraph/engine/GraphEngine.scala` - Graph algorithms
- `knowledgegraph/storage/JsonGraphStore.scala` - JSON persistence
- Test files for all components

---

## Related issue

Closes #217 

This implements the first 3 components from the original issue:
- ✅ LLM-Powered Knowledge Graph Generator
- ✅ Graph Engine & Traversal System  
- ✅ Persistent Storage
- ⏭️ Interactive Frontend (saved for future PR to keep this focused)

---

### Manual Testing
Created and ran `ManualGraphTest.scala` which verified:
- Graph construction with 6 nodes and 9 edges
- Query utilities (findByLabel, findByProperty, getNeighbors)
- BFS and DFS traversal algorithms
- Pathfinding (shortest path, all paths, distance queries)
- JSON save/load with data integrity verification
- Edge cases (missing nodes, invalid paths)

All features worked correctly.

### Unit Tests
- Unit test files created for all components
- Tests don't currently run due to project's `-Werror` flag treating unused import warnings as errors
- This is a build configuration issue, not a code quality issue
- All source code compiles successfully with `sbt "project core" compile`

---

## Checklist

- [x] I have read the [Contributing Guide](https://llm4s.org/reference/contributing)
- [x] PR is small and focused — one change, one reason
- [x] `sbt scalafmtAll` — code is formatted
- [ ] `sbt +test` — tests pass on both Scala 2.13 and 3.x *(tests written but blocked by `-Werror` build flag)*
- [x] New code includes tests
- [x] No unrelated changes included (branched from `main`, not from another PR)
- [x] Commit messages explain the "why"

### CODE REVIEW

**Comment by dhairya-pandya** [comment]
Hello! @vim89 @rorygraves @kannupriyakalra While browsing, I came across this old problem and thought, "Why not tackle this?" Understanding 
These days, graphs are very helpful for GraphRAG, so it seemed worthwhile to implement.
The essential features, such as entity extraction, graph traversal, and storage. 
To keep this PR manageable, the frontend was saved for later.

Just so you know, tests are written, but the `-Werror` build flag prevents them from running, it will continue to give error that particular imports are not utilised at various places. I manually tested it, and everything compiles. Please review this and give feedback on this solution for this issue, wouldlike to improve it if there is any kind of vulnerability in this implementation.

**Comment by dhairya-pandya** [comment]
@vim89 @rorygraves Please review this PR, I have fixed the `-Werror` which was being caused because of unused import statements and tests for some of the functions were not written, so even added that. Please have a look at it & provide feedback where it can be improved.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    } else {
+      Try {
+        val content = new String(Files.readAllBytes(path), StandardCharsets.UTF_8)
+        val json    = ujson.read(content)
+
+        val nodes = json("nodes").arr
+          .map { n =>
+            val id    = n("id").str
+            val label = n("label").str
+            val props = n("properties").obj.map { case (k, v) => k -> v.str }.toMap
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala]: `load()` currently assumes `properties` always exists for nodes and edges. Please default to empty map when field is absent, same as extractor behavior, so persisted and extracted graphs follow one contract.
Guideline principle: API/data contracts should be consistent across module boundaries.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/engine/GraphEngine.scala [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            .filter(_.source == currentNodeId)
+            .flatMap { edge =>
+              val targetNode = graph.nodes.get(edge.target)
+              targetNode match {
+                case Some(node) if !visited.contains(node.id) && filter(node, edge) =>
+                  Some((node.id, depth + 1))
+                case _ => None
+              }
+            }
+          queue = queue ++ neighbors
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/engine/GraphEngine.scala]: Using `List` as BFS queue with repeated append (`queue ++ neighbors`) makes traversal degrade on larger graphs. Consider immutable Queue or deque style structure.
Guideline principle: avoid avoidable allocation and quadratic operations in traversal hot paths.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/item/KnowledgeGraphGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      if (!json.obj.contains("nodes") || !json.obj.contains("edges")) {
+        throw new IllegalArgumentException("JSON must contain 'nodes' and 'edges' fields")
+      }
+
+      val nodes = json("nodes").arr
+        .map { n =>
+          val id    = n("id").str
+          val label = n("label").str
+          val props = if (n.obj.contains("properties")) {
+            n("properties").obj.map { case (k, v) => k -> v.toString.stripPrefix("\"").stripSuffix("\"") }.toMap
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/item/KnowledgeGraphGenerator.scala]: Property parsing here uses `toString` plus quote stripping, which can corrupt values and mishandle non-string JSON types. Please parse `ujson.Value` explicitly (str/num/bool/null/object/arr) before mapping to string.
Guideline principle: parsing external payloads must preserve semantics, not rely on string hacks.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/KnowledgeGraph.scala [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+ * Represents a node (entity) in the knowledge graph.
+ *
+ * @param id Unique identifier for the node
+ * @param label The type or category of the entity (e.g., "Person", "Organization")
+ * @param properties Additional attributes of the entity
+ */
+case class Node(
+  id: String,
+  label: String,
+  properties: Map[String, String] = Map.empty
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/KnowledgeGraph.scala]: High-level design concern: `Map[String, String]` for properties forces lossy coercion and limits future graph use cases. Consider `Map[String, ujson.Value]` or a typed property ADT so extraction and storage do not destroy type information.
Guideline principle: public type design should preserve domain semantics and remain open for extension.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/KnowledgeGraph.scala [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+)
+
+/**
+ * Represents a Knowledge Graph containing nodes and edges.
+ *
+ * @param nodes Map of Node ID to Node object
+ * @param edges List of Edges in the graph
+ */
+case class Graph(
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/knowledgegraph/KnowledgeGraph.scala:40]: Please add a graph integrity check (for example, edge endpoints must exist in node set) at module boundaries like extraction and load. Right now invalid graphs can be created and propagated silently.
Guideline principle: enforce invariants at boundaries, not only in consumers.

**Comment by vim89** [comment]
Adding one extra design-focused follow-up from second pass: low-level parsing and queue complexity are already flagged; at high-level, property typing and graph integrity invariants should also be hardened to make this module stable for production-grade Graph RAG use. I did not run tests or build locally.

**Comment by Vi-shub** [comment]
Hi @dhairya-pandya this is good initiative to add graph RAG but I would advice you to do this in phases, please do not push it in single PR. Also have detailed architecture how it would work with existing modules and functions. Make issue first before directly opening PR and have discussion there. There is a channel in discord for discussions.

**Comment by dhairya-pandya** [comment]
@Vi-shub This PR solves the issue #217, and definitely this is a big PR, I have not even solved the interface part of it, but was thinking to do that in a seperate PR. Still should I break down this PR or do the changes as per reviews by @vim89. As I am taking my time and solving this issue, should I further breakdown this PR by opening a new issue and discussing the architecture to integrate it with existing infra? Thanks for reviewing it! @vim89 @Vi-shub

**Comment by dhairya-pandya** [comment]
I think I should breakdown this PR into multiple smaller PRs, firstly closing this PR and starting with implementing core data structures in the first PR, and then doing it in phases further on! Would like to know your views on this @vim89 @Vi-shub

**Comment by rorygraves** [comment]
> @Vi-shub This PR solves the issue #217, and definitely this is a big PR, I have not even solved the interface part of it, but was thinking to do that in a seperate PR. Still should I break down this PR or do the changes as per reviews by @vim89. As I am taking my time and solving this issue, should I further breakdown this PR by opening a new issue and discussing the architecture to integrate it with existing infra? Thanks for reviewing it! @vim89 @Vi-shub

This is a great, start, but it does not solve the issue in 217 - an in memory  json storage based knowledge graph is an interesting toy - but not a production ready utility.  It would need persistent storage, documenation, samples, tools and much more.

File: modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala (hunk 1 above)
Comment (dhairya-pandya) [reply modules/core/src/main/scala/org/llm4s/knowledgegraph/storage/JsonGraphStore.scala]: Fixed it! The load() method now explicitly checks for the existence of the properties field and defaults to Map.empty[String, ujson.Value] if it is missing.

**Comment by dhairya-pandya** [comment]
@rorygraves @vim89 Thank you both for the detailed and critical reviews!

I initially considered splitting this PR as @Vi-shub suggested, but after Rory's review, I decided to lend extra time and effort to implement the core module as a whole at this point. I wanted to ensure the data structures, extraction logic, and storage all adhered to a consistent, hardened contract from day one rather than deferring the "correctness gaps" to future PRs.

I've pushed a major update that squashes all commits and addresses **every point** from your reviews:

### Fixes Implemented
1.  **Type Safety**: Switched `properties` to `Map[String, ujson.Value]` to preserve original JSON types (no more string stripping hacks).
2.  **Robustness**: `JsonGraphStore` now explicitly handles missing `properties` fields with `Map.empty` defaults.
3.  **Stability**: Replaced unsafe map access in `GraphEngine` with safe `Option` handling.
4.  **Performance**: Refactored BFS and Shortest Path to use `scala.collection.immutable.Queue` for O(1) operations, avoiding the O(n) list append penalty.
5.  **Integrity**: Added strict `graph.validate()` method to ensure no dangling edges exist at module boundaries (extraction/load).
6.  **Testing**: Expanded `JsonGraphStoreTest` to cover malformed JSON and edge cases.
7.  **Refactoring**: Renamed `item` package to `extraction`.
8.  **Determinism**: Set `temperature = 0.0` for LLM extraction.

The PR is now a single clean commit. Please re-review and provide feedback regarding the same! @vim89 @rorygraves
